presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -484,6 +484,20 @@ public PlanWithProperties visitSort(SortNode node, Context context)
                         gatheringExchange(idAllocator.getNextId(), REMOTE, child.getNode()),
                         child.getProperties());
             }
+            else {
+                // current plan so far is single node, so local properties are effectively global properties
+                // skip the SortNode if the local properties guarantee ordering on Sort keys
+                // TODO: This should be extracted as a separate optimizer once the planner is able to reason about the ordering of each operator
+                List<LocalProperty<Symbol>> desiredProperties = new ArrayList<>();
+                for (Symbol symbol : node.getOrderBy()) {
+                    desiredProperties.add(new SortingProperty<>(symbol, node.getOrderings().get(symbol)));
+                }
+
+                if (LocalProperties.match(child.getProperties().getLocalProperties(), desiredProperties).stream()
+                        .noneMatch(Optional::isPresent)) {
+                    return child;
+                }
+            }
```

Comment:
I'd rather not go down that path. The proper way to address this is to add "traits" to the new optimizer (effective predicate, sortedness, partitioning/distribution, uniqueness, nullability, type, etc.) that can be derived bottom up and memoized. Rules would then have access to these, just like they have access to the plan structure to make decisions.